src/Microsoft.IdentityModel.Clients.ActiveDirectory/Internal/Instance/Authenticator.cs

```diff
@@ -90,26 +88,29 @@ public Authenticator(string authority, bool validateAuthority)
 
         public string SelfSignedJwtAudience { get; private set; }
 
+        public string PreferredCache { get; private set; }
+
+        public string[] Aliases { get; private set; }
+
         public Guid CorrelationId { get; set; }
 
         public async Task UpdateFromTemplateAsync(CallState callState)
         {
             if (!this.updatedFromTemplate)
             {
                 var authorityUri = new Uri(this.Authority);
-                string host = authorityUri.Authority;
                 string path = authorityUri.AbsolutePath.Substring(1);
                 string tenant = path.Substring(0, path.IndexOf("/", StringComparison.Ordinal));
-
-                AuthenticatorTemplate matchingTemplate = await AuthenticatorTemplateList.FindMatchingItemAsync(this.ValidateAuthority, host, tenant, callState).ConfigureAwait(false);
-
-                this.AuthorizationUri = matchingTemplate.AuthorizeEndpoint.Replace("{tenant}", tenant);
-                this.DeviceCodeUri = matchingTemplate.DeviceCodeEndpoint.Replace("{tenant}", tenant);
-                this.TokenUri = matchingTemplate.TokenEndpoint.Replace("{tenant}", tenant);
-                this.UserRealmUri = CanonicalizeUri(matchingTemplate.UserRealmEndpoint);
+                var entry = await InstanceDiscovery.GetMetadataEntry(authorityUri.Host, this.ValidateAuthority, callState);
+                this.AuthorizationUri = InstanceDiscovery.AuthorizeEndpointTemplate.Replace("{host}", entry.PreferredNetwork).Replace("{tenant}", tenant);
```

Comment:
you can use String.Format instead of replace.

Reply:
Thanks for the hint but, after some experiment, I think the current template is more readable with meaningful placeholders such as `https://{host}/{tenant}/oauth2/authorize`, rather than `https://{0}/{1}/oauth2/authorize`.

C# 6.0 supports [Interpolated Strings](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/interpolated-strings), but it requires us to define the corresponding local variables. Not sure if we really like that better.

Reply:
OK then. The next commit will go with the interpolated strings approach in all other lines, but not this particular line. Because this line uses a [string variable, which can not be used with string interpolation](https://stackoverflow.com/questions/43288857/string-interpolation-with-variable-content-in-c-sharp).

Reply:
I'd recommend instead of having an AuthorizeEndpointTemplate property, create a FormatAuthorizeEndpoint(string host, string tenant) function that would return the fully formatted string. Is there a reason this needs to be a property?